homeassistant/components/thread/dataset_store.py

```
@@ -0,0 +1,150 @@
+        dataset = tlv_parser.parse_tlv(self.tlv)
+        return dataset.get(tlv_parser.MeshcopTLVType.NETWORKNAME)
+
+    @property
+    def pan_id(self) -> str | None:
+        """Return PAN ID as a hex string."""
+        dataset = tlv_parser.parse_tlv(self.tlv)
+        return dataset.get(tlv_parser.MeshcopTLVType.PANID)
+
+    def to_json(self) -> dict[str, Any]:
+        """Return a JSON serializable representation for storage."""
+        return {
+            "created": self.created.isoformat(),
+            "id": self.id,
+            "preferred": self.preferred,
+            "source": self.source,
+            "tlv": self.tlv,
+        }
+
+
+class DatasetStore:
+    """Class to hold a collection of thread datasets."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the dataset store."""
+        self.hass = hass
+        self.datasets: dict[str, DatasetEntry] = {}
+        self._store: Store[dict[str, list[dict[str, Any]]]] = Store(
+            hass,
+            STORAGE_VERSION_MAJOR,
+            STORAGE_KEY,
+            atomic_writes=True,
+            minor_version=STORAGE_VERSION_MINOR,
+        )
+
+    @callback
+    def async_add(self, source: str, tlv: str) -> None:
+        """Add dataset, does nothing if it already exists."""
+        # Make sure the tlv is valid
+        tlv_parser.parse_tlv(tlv)
+        # Bail out if the dataset already exists
+        if any(dataset for dataset in self.datasets.values() if dataset.tlv == tlv):
+            return
+
+        # Set to preferred if this is the first dataset
+        preferred = not bool(self.datasets)
+        dataset = DatasetEntry(preferred=preferred, source=source, tlv=tlv)
+        self.datasets[dataset.id] = dataset
+        self.async_schedule_save()
+
+    @callback
+    def async_delete(self, dataset_id: str) -> None:
+        """Delete dataset."""
+        if not (dataset := self.datasets.get(dataset_id)):
```

let's raise an error instead of silently fail.